pkg/engine/lifecycletest/fuzzing/provider.go

```diff
@@ -0,0 +1,562 @@
+}
+
+// Returns a DiffF-compatible callback that implements this ProviderDiffSpec.
+func (pds ProviderDiffSpec) AsDiffF() func(context.Context, plugin.DiffRequest) (plugin.DiffResponse, error) {
+	return func(_ context.Context, req plugin.DiffRequest) (plugin.DiffResponse, error) {
+		if action, ok := pds[req.URN]; ok {
+			switch action {
+			case ProviderDiffDeleteBeforeReplace:
+				return plugin.DiffResponse{
+					Changes:             plugin.DiffSome,
+					ReplaceKeys:         []resource.PropertyKey{"__replace"},
+					DeleteBeforeReplace: true,
+				}, nil
+			case ProviderDiffDeleteAfterReplace:
+				return plugin.DiffResponse{
+					Changes:             plugin.DiffSome,
+					ReplaceKeys:         []resource.PropertyKey{"__replace"},
+					DeleteBeforeReplace: false,
+				}, nil
+			case ProviderDiffChange:
+				return plugin.DiffResponse{Changes: plugin.DiffSome}, nil
+			case ProviderDiffFailure:
+				return plugin.DiffResponse{}, fmt.Errorf("diff failure for %s", req.URN)
+			}
+		}
+
+		return plugin.DiffResponse{}, nil
+	}
+}
+
+// A ProviderReadSpec specifies the behavior of a provider's read function. It maps resource URNs to the action that
+// should be taken if Read is called on that URN. The absence of a URN in the map indicates that the default behavior (a
+// successful read that reports that the resource exists) should be taken.
+type ProviderReadSpec map[resource.URN]ProviderReadSpecAction
+
+// ProviderReadSpecAction captures the set of actions that can be taken by a Read implementation for a given resource.
+type ProviderReadSpecAction string
+
+const (
+	// Return a result that indicates that the resource has been deleted.
+	ProviderReadDeleted ProviderReadSpecAction = "provider.read-deleted"
+	// Fail the Read operation.
+	ProviderReadFailure ProviderReadSpecAction = "provider.read-failure"
+)
+
+// Implements PrettySpec.Pretty. Returns a human-readable string representation of this ProviderReadSpec, suitable for
+// use in debugging output or error messages.
+func (prs ProviderReadSpec) Pretty(indent string) string {
+	if len(prs) == 0 {
+		return ""
+	}
+
+	rendered := fmt.Sprintf("\n%sRead:", indent)
+	for r, action := range prs {
+		switch action {
+		case ProviderReadDeleted:
+			rendered += fmt.Sprintf("\n%s  -  %s", indent, Colored(r))
+		case ProviderReadFailure:
+			rendered += fmt.Sprintf("\n%s  !  %s", indent, Colored(r))
+		}
+	}
+
+	return rendered
+}
+
+// Returns a ReadF-compatible callback that implements this ProviderReadSpec.
+func (prs ProviderReadSpec) AsReadF() func(context.Context, plugin.ReadRequest) (plugin.ReadResponse, error) {
+	return func(_ context.Context, req plugin.ReadRequest) (plugin.ReadResponse, error) {
+		if action, ok := prs[req.URN]; ok {
+			switch action {
+			case ProviderReadDeleted:
+				return plugin.ReadResponse{}, nil
+			case ProviderReadFailure:
+				return plugin.ReadResponse{Status: resource.StatusPartialFailure}, fmt.Errorf("read failure for %s", req.URN)
```

Comment:
Why is a read returning StatusPartialFailure vs StatusUnknown we have in other cases? It's just what's done for read?

Reply:
This is a good call-out. I've taken this from a case we had in a previous lifecycle test, but I'm not sure it matters/is required. I'm going to make it `StatusUnknown` and expect it will still work as intended. Thanks for spotting!